### ISSUE TITLE
git snapshot compilation fails on FreeBSD due to C++ changes

### ISSUE THREAD

**Issue by czanik** [syslog-ng/syslog-ng#4496]
While trying to do my weekly git snapshot builds I ran into a compile error on FreeBSD, due to https://github.com/syslog-ng/syslog-ng/pull/4484

```
/bin/sh ./libtool  --tag=CXX   --mode=compile c++ -DHAVE_CONFIG_H -I.  -I./lib -I./modules -I./lib -I./modules -I/usr/local/include/glib-2.0 -I/usr/local/lib/glib-2.0/include -I/usr/local/include -pthread  -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_32 -I./lib/eventlog/src -I./lib/eventlog/src -I/usr/local/include  -I/usr/include   -isystem ./lib/ivykis/src/include -isystem ./lib/ivykis/src/include  -D_GNU_SOURCE -D_DEFAULT_SOURCE -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64  -I./modules/examples/sources/random-choice-generator -I./modules/examples/sources/random-choice-generator -O2 -pipe -fstack-protector-strong -fno-strict-aliasing  -MT modules/examples/sources/random-choice-generator/librandom_choice_generator_cpp_la-random-choice-generator.lo -MD -MP -MF modules/examples/sources/random-choice-generator/.deps/librandom_choice_generator_cpp_la-random-choice-generator.Tpo -c -o modules/examples/sources/random-choice-generator/librandom_choice_generator_cpp_la-random-choice-generator.lo `test -f 'modules/examples/sources/random-choice-generator/random-choice-generator.cpp' || echo './'`modules/examples/sources/random-choice-generator/random-choice-generator.cpp
libtool: compile:  c++ -DHAVE_CONFIG_H -I. -I./lib -I./modules -I./lib -I./modules -I/usr/local/include/glib-2.0 -I/usr/local/lib/glib-2.0/include -I/usr/local/include -pthread -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_32 -I./lib/eventlog/src -I./lib/eventlog/src -I/usr/local/include -I/usr/include -isystem ./lib/ivykis/src/include -isystem ./lib/ivykis/src/include -D_GNU_SOURCE -D_DEFAULT_SOURCE -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I./modules/examples/sources/random-choice-generator -I./modules/examples/sources/random-choice-generator -O2 -pipe -fstack-protector-strong -fno-strict-aliasing -MT modules/examples/sources/random-choice-generator/librandom_choice_generator_cpp_la-random-choice-generator.lo -MD -MP -MF modules/examples/sources/random-choice-generator/.deps/librandom_choice_generator_cpp_la-random-choice-generator.Tpo -c modules/examples/sources/random-choice-generator/random-choice-generator.cpp  -fPIC -DPIC -o modules/examples/sources/random-choice-generator/.libs/librandom_choice_generator_cpp_la-random-choice-generator.o
In file included from modules/examples/sources/random-choice-generator/random-choice-generator.cpp:25:
In file included from ./modules/examples/sources/random-choice-generator/random-choice-generator.h:28:
In file included from ./lib/driver.h:29:
In file included from ./lib/logpipe.h:30:
In file included from ./lib/cfg.h:38:
In file included from ./lib/stats/stats.h:29:
In file included from ./lib/stats/stats-cluster.h:27:
In file included from ./lib/stats/stats-counter.h:28:
./lib/atomic-gssize.h:118:10: error: cannot initialize a parameter of type 'gssize *' (aka 'long *') with an rvalue of type 'void *'
  return g_atomic_pointer_compare_and_exchange(&a->value, oldval, newval);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/glib-2.0/glib/gatomic.h:258:44: note: expanded from macro 'g_atomic_pointer_compare_and_exchange'
    __atomic_compare_exchange_n ((atomic), (void *) (&(gapcae_oldval)), (newval), FALSE, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST) ? TRUE : FALSE; \
                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
gmake[4]: *** [Makefile:25664: modules/examples/sources/random-choice-generator/librandom_choice_generator_cpp_la-random-choice-generator.lo] Error 1
gmake[3]: *** [Makefile:26755: all-recursive] Error 1
gmake[2]: *** [Makefile:11471: all] Error 2
gmake[2]: Leaving directory '/root/freebsd/syslog-ng4-devel/work/syslog-ng-4.2.0.38.g0a39c11'
*** Error code 1

```

**Comment by czanik** [comment]
```--disable-cpp``` can be used as a workaround.

**Comment by czanik** [comment]
@HofiOne mentioned today, that the C++ plugin has problems with CLANG, and only works with GCC. FreeBSD only has CLANG available by default. GCC is a huge dependency people only install on FreeBSD when nothing else works. For now ```--disable-cpp``` will be used in the syslog-ng port. Of course, I'll change it as soon as there is more than just the example module.

**Comment by alltilla** [comment]
You don't need to install gcc to run the C++ modules, it is only needed during building/packaging. I think if gcc is available on FreeBSD, you should install it during packaging as a build dependency only, so users get the C++ based features.

The next real C++ based module will be OpenTelemetry. I checked centos-7 and the protoc grpc plugin, which is necessary for build time, is not available there. I have not checked all the RPM based OSs, though. Maybe someone should package it sometime. 😊

**Comment by Schievel1** [comment]
Compiling gcc just for one small plugin of a small package isn't really what users usually want.
`--disable-cpp` disables those modules as well, doesn't it?

https://bugs.gentoo.org/914645

**Comment by czanik** [comment]
Yes, ```--disable-cpp``` disables compiling these modules. However, there were FreeBSD users asking for OpenTelemetry support, so I checked it again:

 * using clang still does not work: https://github.com/syslog-ng/syslog-ng/issues/4652
 * gcc works, but modules do not load: https://github.com/syslog-ng/syslog-ng/issues/4653

I even did the packaging part locally, but obviously it's not yet submitted. In FreeBSD ports there are a list of configuration options. The default options are enabled in packages, c++-based are not among default options. I configured the ports Makefile in a way, that clang is used by default, and gcc is pulled in only if c++ modules are enabled. A kind of far from perfect, better than nothing :-) Being able to use clang for everything would be better. But, as a first step, getting gcc compiled modules working would also be great ;-)
